src/Workspaces/Core/Portable/ExternalAccess/UnitTesting/UnitTestingIncrementalAnalyzerProvider.cs

```diff
@@ -10,11 +11,26 @@ namespace Microsoft.CodeAnalysis.ExternalAccess.UnitTesting
     internal class UnitTestingIncrementalAnalyzerProvider : IIncrementalAnalyzerProvider
     {
         private readonly IUnitTestingIncrementalAnalyzerProviderImplementation _incrementalAnalyzerProvider;
+        private IIncrementalAnalyzer _analyzer;
+        private Workspace _workspace;
 
         public UnitTestingIncrementalAnalyzerProvider(IUnitTestingIncrementalAnalyzerProviderImplementation incrementalAnalyzerProvider)
             => _incrementalAnalyzerProvider = incrementalAnalyzerProvider;
 
         public IIncrementalAnalyzer CreateIncrementalAnalyzer(Workspace workspace)
-            => new UnitTestingIncrementalAnalyzer(_incrementalAnalyzerProvider.CreateIncrementalAnalyzer(workspace));
+        {
+            if (_workspace != null && _workspace != workspace)
+            {
+                throw new ArgumentException(nameof(workspace));
+            }
+
+            if (_analyzer == null)
```

Comment:
Should we throw in the else case (i.e. if CreateIncrementalAnalyzer ends up being called twice) similar to what we do in UnitTestingSolutionCrawlerServiceAccessor.AddAnalyzerProvider() above?

Reply:
as stated above, this ends up being called multiple times because of the Reanalyze method. we can push the caching to the caller class if it makes things more clear

Reply:
I think that given that we pass in the analyzer provider to solution crawler, it will be safer if the analyzer provider caches the incremental analyzer instance

Reply:
Perhaps adding some comments around what is happening here (or atop _analyzer field) will make things clearer.